torch/csrc/distributed/c10d/ProcessGroupUCC.cpp

```diff
@@ -0,0 +1,1575 @@
+  }
+  torch_ucc_config.blocking_wait[(std::uint8_t)OpType::ALLGATHER] =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_ALLGATHER_BLOCKING_WAIT"));
+  torch_ucc_config.blocking_wait[(std::uint8_t)OpType::ALLREDUCE] =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_ALLREDUCE_BLOCKING_WAIT"));
+  torch_ucc_config.blocking_wait[(std::uint8_t)OpType::ALLTOALL_BASE] =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_ALLTOALL_BLOCKING_WAIT"));
+  torch_ucc_config.blocking_wait[(std::uint8_t)OpType::BROADCAST] =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_BCAST_BLOCKING_WAIT"));
+  torch_ucc_config.blocking_wait[(std::uint8_t)OpType::REDUCE_SCATTER] =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_REDUCE_SCATTER_BLOCKING_WAIT"));
+  torch_ucc_config.blocking_wait[(std::uint8_t)OpType::SCATTER] =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_SCATTER_BLOCKING_WAIT"));
+  torch_ucc_config.blocking_wait[(std::uint8_t)OpType::GATHER] =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_GATHER_BLOCKING_WAIT"));
+  torch_ucc_config.use_future =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_USE_FUTURE"));
+  torch_ucc_config.enable_profiling =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_PROFILING_ENABLE"));
+  torch_ucc_config.shared_comm =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_SHARED_COMM"));
+  torch_ucc_config.use_allgatherv =
+      std::stoi(torch_ucc_envs_map.at("TORCH_UCC_USE_ALLGATHERV"));
+}
+
+void check_device(c10::Device dev1, c10::Device dev2) {
+  if (dev1.is_cuda() && dev2.is_cuda() && dev1 != dev2) {
+    throw std::runtime_error("ProcessGroupUCC multidevice is not supported");
+  }
+}
+
+void check_tensor(const std::vector<at::Tensor>& tensors) {
+  if (tensors.size() != 1) {
+    throw std::runtime_error("ProcessGroupUCC takes 1 tensor");
+  }
+  if (!tensors[0].is_contiguous()) {
+    throw std::runtime_error(
+        "ProcessGroupUCC input tensor has to be contiguous");
+  }
+  if (tensors[0].is_sparse()) {
+    throw std::runtime_error("ProcessGroupUCC input tensor has to be dense");
+  }
+  // TODO: check cuda case
+}
+
+ProcessGroupUCC::WorkUCC::~WorkUCC() {
+#ifdef USE_CUDA
+  if (fence && ep) {
+    std::lock_guard<std::mutex> lock(ep->event_pool_mutex);
+    ep->event_pool.push(std::move(fence));
+  }
+#endif
+}
+
+void ProcessGroupUCC::WorkUCC::setException() {
+  if (exception() || !entry_) {
+    return;
+  }
+  exception_ = entry_->eptr_;
+}
+
+void ProcessGroupUCC::WorkUCC::setAndThrowException() {
+  setException();
+  if (exception()) {
+    std::rethrow_exception(exception());
+  }
+}
+
+bool ProcessGroupUCC::WorkUCC::isCompleted() {
+  if (!entry_) {
+    return true;
+  }
+  setException();
+  return exception() || entry_->status_ <= 0;
+}
+
+bool ProcessGroupUCC::WorkUCC::isSuccess() const {
+  if (!entry_) {
+    return true;
+  }
+  return !exception() && entry_->status_ == 0;
+}
+
+bool ProcessGroupUCC::WorkUCC::wait(std::chrono::milliseconds /* unused */) {
+#ifdef USE_CUDA
```

Comment:
I think `USE_CUDA` is always true for now? But if you do want to use UCC on CPU-only builds in the future, we can keep this guard for now, so that in the future, we can easily figure out which code to move to `libtorch_cuda.so`.

Reply:
let's keep it for CPU support as you mentioned.